efficientdet/keras/train.py

```diff
@@ -0,0 +1,214 @@
+    raise RuntimeError('You must specify --training_file_pattern for training.')
+  if FLAGS.mode in ('eval', 'train_and_eval'):
+    if FLAGS.validation_file_pattern is None:
+      raise RuntimeError('You must specify --validation_file_pattern '
+                         'for evaluation.')
+
+  params = dict(config.as_dict(),
+                model_name=FLAGS.model_name,
+                iterations_per_loop=FLAGS.iterations_per_loop,
+                model_dir=FLAGS.model_dir,
+                num_examples_per_epoch=FLAGS.num_examples_per_epoch,
+                strategy=FLAGS.strategy,
+                batch_size=FLAGS.train_batch_size,
+                eval_batch_size=FLAGS.eval_batch_size,
+                num_shards=FLAGS.num_cores,
+                val_json_file=FLAGS.val_json_file,
+                testdev_dir=FLAGS.testdev_dir,
+                mode=FLAGS.mode)
+
+  # set mixed precision policy by keras api which will generate mixed precision model
+  precision = utils.get_precision(params['strategy'], params['mixed_precision'])
+  policy = tf.keras.mixed_precision.experimental.Policy(precision)
+  tf.keras.mixed_precision.experimental.set_policy(policy)
+
+  def get_dataset(is_training, params):
+    global_batch_size = FLAGS.train_batch_size if is_training else FLAGS.eval_batch_size
+    batch_size = global_batch_size // ds_strategy.num_replicas_in_sync
+    file_pattern = FLAGS.training_file_pattern if is_training else FLAGS.validation_file_pattern
+    params = copy.deepcopy(params)
+    update_params = dict(batch_size=batch_size) if is_training else dict(
+        eval_batch_size=batch_size)
+    params.update(update_params)
+    return dataloader.InputReader(
+        file_pattern,
+        is_training=is_training,
+        use_fake_data=FLAGS.use_fake_data,
+        max_instances_per_image=config.max_instances_per_image)(params)
+
+  with ds_strategy.scope():
+    model = EfficientDetNetTrain(params['model_name'], config)
+  ckpt_path = tf.train.latest_checkpoint(FLAGS.model_dir)
+  if ckpt_path:
+    model.load_weights(ckpt_path)
+  model.freeze_vars(params['var_freeze_expr'])
+  with ds_strategy.scope():
+    model.compile(optimizer=get_optimizer(params),
+                  loss={
+                      "box_loss":
+                          BoxLoss(params['delta'],
+                                  reduction=tf.keras.losses.Reduction.NONE),
+                      "box_iou_loss":
+                          BoxIouLoss(params['iou_loss_type'],
+                                     reduction=tf.keras.losses.Reduction.NONE),
+                      "class_loss":
+                          FocalLoss(params['alpha'],
+                                    params['gamma'],
+                                    label_smoothing=params['label_smoothing'],
+                                    reduction=tf.keras.losses.Reduction.NONE)
+                  })
+    model.train(get_callbacks(params, FLAGS.profile),
```

Comment:
The keras doc recommends  "model.fit" instead of train. Should we use model.fit?

Reply:
```model.fit``` process input by nest.flatten and zip, which only return top10 values in labels dict.

Reply:
Maybe we should convert the labels into a list as well?
I roughly read the fit function: looks like it does a lot of extra work for multi-worker and multi-GPU/TPU training. It might be difficult to capture all of them in the customized train. WDYT?

Reply:
It could be done in next PR.
In fact, I have used the custom train loop in my repo for half a year.

Reply:
Aha, okay, then let's first try to use this customized train.